### PR TITLE
Null pointer protection

### PR DESCRIPTION
**Pull request by divyakoshy**
**Description**

Data received from an external source should be null-checked before being used; preventing crashes due to null-pointer errors. These checks must be done on both input data and output data.

**Design**

All `grpc` services will be updated to check for null fields before processing the request. A new `IsEmpty` method will be added to all objects that must be tested further.

### CODE REVIEW

File: contract/config.go [hunk 1]
```diff
@@ -28,3 +28,11 @@ type Config struct {
 	RenExBalancesAddress    string  `json:"renExBalancesAddress"`
 	RenExSettlementAddress  string  `json:"renExSettlementAddress"`
 }
+
+// IsEmpty returns true if Config or any of its fields are nil.
+func (config *Config) IsEmpty() bool {
```
Comment (loongy) [inline contract/config.go]: A more appropriate name for this method is `IsNil` or `IsPartiallyNil`.

File: grpc/conn.go [hunk 2]
```diff
@@ -15,6 +15,9 @@ import (
 // call grpc.ClientConn.Close to terminate all the pending operations after
 // this function returns.
 func Dial(ctx context.Context, multiAddress identity.MultiAddress) (*grpc.ClientConn, error) {
+	if multiAddress.IsEmpty() {
```
Comment (loongy) [inline grpc/conn.go]: Likewise, this method should probably be called `IsNil`.

File: grpc/orderbook.go [hunk 3]
```diff
@@ -23,6 +24,9 @@ func NewOrderbookClient() orderbook.Client {
 
 // OpenOrder implements the orderbook.Client interface.
 func (client *orderbookClient) OpenOrder(ctx context.Context, multiAddr identity.MultiAddress, orderFragment order.EncryptedFragment) error {
+	if orderFragment.IsEmpty() {
```
Comment (loongy) [inline grpc/orderbook.go]: Again, in general consider the naming of these methods to be more representative of the internal check — `IsNil`.

File: grpc/orderbook.go [hunk 4]
```diff
@@ -23,6 +24,9 @@ func NewOrderbookClient() orderbook.Client {
...
+		return fmt.Errorf("cannot open order for empty order fragment")
```
Comment (loongy) [inline grpc/orderbook.go]: More consider error descriptions. If the error does not need to be parameterised, then declare the error as a value and use it.

```go
var ErrOrderFragmentIsNil = errors.New("order fragment is nil")
```

File: grpc/orderbook.go [hunk 5]
```diff
@@ -56,12 +60,19 @@ func NewOrderbookService(server orderbook.Server) OrderbookService {
 
 // Register implements the Service interface.
 func (service *OrderbookService) Register(server *Server) {
+	if server == nil {
+		logger.Network(logger.LevelError, fmt.Sprint("cannot register with invalid server"))
```
Comment (loongy) [inline grpc/orderbook.go]: Why are you using `fmt.Sprintf`?

File: grpc/swarm.go [hunk 6]
```diff
@@ -163,6 +177,11 @@ func (service *SwarmService) Register(server *Server) {
 // signed identity.MultiAddress of the client it will return its own signed
 // identity.MultiAddress in a PingResponse.
 func (service *SwarmService) Ping(ctx context.Context, request *PingRequest) (*PingResponse, error) {
+	// Check for empty or invalid request fields.
+	if request == nil || request.MultiAddress == nil {
+		return nil, ErrInvalidRequest
```
Comment (loongy) [inline grpc/swarm.go]: Using something like `ErrInvalidRequest` either needs to be parameterised, or you should create a more specific error value, e.g. `ErrMultiAddressIsNil`. Having only "invalid request" as a message does not tell us what went wrong during debugging.

File: grpc/swarm.go [hunk 7]
```diff
@@ -217,6 +239,10 @@ func (service *SwarmService) Pong(ctx context.Context, request *PongRequest) (*P
 // responsibility to its swarm.Server to return identity.MultiAddresses that
 // are close to the queried identity.Address.
 func (service *SwarmService) Query(ctx context.Context, request *QueryRequest) (*QueryResponse, error) {
+	if request == nil || len(request.Address) == 0 {
+		return nil, ErrInvalidRequest
```
Comment (loongy) [inline grpc/swarm.go]: Same here, consider something like `ErrAddressIsNil`.

File: identity/multiaddress.go [hunk 8]
```diff
@@ -154,6 +154,11 @@ func (multiAddress *MultiAddress) UnmarshalStringJSON(data []byte) error {
 	return nil
 }
 
+// IsEmpty checks if the multiAddress is nil.
+func (multiAddress *MultiAddress) IsEmpty() bool {
+	return (multiAddress == nil || multiAddress.baseMultiAddress == nil)
```
Comment (loongy) [inline identity/multiaddress.go]: What about the non-base address? Strings can be checked against the empty string `""`.

File: orderbook/orderbook.go [hunk 9]
```diff
@@ -93,6 +93,10 @@ func NewOrderbook(addr identity.Address, rsaKey crypto.RsaKey, pointerStore Poin
 
 // OpenOrder implements the Server interface.
 func (orderbook *orderbook) OpenOrder(ctx context.Context, encryptedOrderFragment order.EncryptedFragment) error {
+	if encryptedOrderFragment.IsEmpty() {
+		return ErrOrderFragmentNotFound
```
Comment (loongy) [inline orderbook/orderbook.go]: Using `ErrOrderFragmentNotFound` is not the right error to be using here.

File: stream/channel.go [hunk 10]
```diff
@@ -12,6 +12,10 @@ import (
 // again with the same server
 var ErrAlreadyRegistered = errors.New("client has already been registered with the server")
 
+// ErrNullMessage is returned when a client tries to send or receive a null
+// message.
+var ErrNullMessage = errors.New("null message")
```
Comment (loongy) [inline stream/channel.go]: In Go, it is `nil` not `null`.

Comment (loongy) [reply stream/channel.go]: For consistency with the other proposed messages, consider `ErrMessageIsNil`.

File: order/fragment.go [hunk 11]
```diff
@@ -245,6 +245,11 @@ func (fragment *EncryptedFragment) Decrypt(privKey *rsa.PrivateKey) (Fragment, e
 	return decryptedFragment, nil
 }
 
+// IsNil checks if an EncryptedFragment is null.
+func (fragment *EncryptedFragment) IsNil() bool {
+	return fragment == nil || fragment.ID == (FragmentID{}) || len(fragment.ID) != 32 || len(fragment.OrderID) != 32
```
Comment (tok-kkk) [inline order/fragment.go]: len(fragment.ID) != 32 and  len(fragment.OrderID) != 32 will always be true as they are [32]byte

File: registry/epoch.go [hunk 12]
```diff
@@ -38,3 +38,8 @@ func (ξ *Epoch) Pod(addr identity.Address) (Pod, error) {
 	}
 	return Pod{}, ErrPodNotFound
 }
+
+// IsNil returns true if the Epoch is nil.
+func (ξ *Epoch) IsNil() bool {
+	return ξ == nil || len(ξ.Hash) != 32 || len(ξ.Darknodes) == 0 || len(ξ.Pods) == 0
```
Comment (tok-kkk) [inline registry/epoch.go]: len(ξ.Hash) != 32 will always be true

**Pull request merged**